### ISSUE TITLE
Integrate BackgroundParser

### ISSUE THREAD

**Issue by Therzok** [mono/monodevelop#4910]
As soon as we get rid of ParsedDocument for C#, we can migrate to Background parser.
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Workspace/BackgroundParser.cs,e1c4fec4b37f8e69

Handles computation of a document's syntax tree when the document is opened/closed (cancelling on close), and other document/workspace states changes

The IDE document should only trigger change notifications.

> VS bug [#623288](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/623288)

**Comment by Therzok** [comment]
Our current document parsing system does not handle one specific scenario that the background parser is optimized for.

We might have some solid improvements if we switch:
https://github.com/dotnet/roslyn/blob/7666140edf112b8efee0eae605a9b44f78bfecdf/src/Features/Core/Portable/Workspace/BackgroundParser.cs#L190

**Comment by Therzok** [comment]
```
// We end up creating a chain of parsing tasks that each attempt to produce 
            // the appropriate syntax tree for any given document. Once we start work to create 
            // the syntax tree for a given document, we don't want to stop. 
            // Otherwise we can end up in the unfortunate scenario where we keep cancelling work, 
            // and then having the next task re-do the work we were just in the middle of. 
            // By not cancelling, we can reuse the useful results of previous tasks when performing later steps in the chain.
            //
            // we still cancel whole task if the task didn't start yet. we just don't cancel if task is started but not finished yet.
```

**Comment by Therzok** [comment]
This has been done in the 8.0 branch.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-40] Comment by **xamarinc**



Fixed in version 8.1.0.1701 (master-vnext)



Pull Request #121 merged by: Aaron Bockover  
 Author: xamarin  
 Commit: 04ae159fc7e8701957b9ad048b5d944afa562a28 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-54] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #121 merged by: Aaron Bockover  
 Author: xamarin  
 Commit: 04ae159fc7e8701957b9ad048b5d944afa562a28 (xamarin/monodevelop)
